pkg/kv/kvserver/batcheval/cmd_end_transaction.go

```diff
@@ -882,45 +882,74 @@ func RunCommitTrigger(
 			"commit wait. Was its timestamp bumped after acquiring latches?", txn, ct.Kind())
 	}
 
+	// Used by both splits and merges.
+	maybeWrapReplicaCorruptionError := func(ctx context.Context, err error) error {
+		if err == nil {
+			log.KvExec.Fatalf(ctx, "unexpected nil error")
+		}
+		if info := pebble.ExtractDataCorruptionInfo(err); info != nil {
+			// Data corruption errors due to external SSTable references getting
+			// deleted should not be wrapped in replica corruption errors. This
+			// ensures that we simply fail the split or merge and propagate the error,
+			// but don't crash the process. In such cases, an excise command should be
+			// used to get out of this data corruption situation.
+			return err
+		}
+		// Otherwise, fail the split or merge with a critical error that crashes the
+		// process. Reporting a replica corruption error ensures this. See
+		// setCorruptRaftMuLocked.
+		return kvpb.MaybeWrapReplicaCorruptionError(ctx, err)
+	}
+
 	// Stage the commit trigger's side-effects so that they will go into effect on
 	// each Replica when the corresponding Raft log entry is applied. Only one
 	// commit trigger can be set.
 	if ct.GetSplitTrigger() != nil {
+		sl := MakeStateLoader(rec)
```

Comment:
nit: might be good to have a `loadSplitTriggerInput` helper, next to `splitTriggerHelperInput` struct. Instead of inlining stuff in otherwise a pretty massive func.